File: scripts/export-lulo-mdb-csv.mjs
/**
 * Exporta tablas LuloWin (.mdb) a CSV sin ODBC (usa mdb-reader, igual que la app).
 *
 * Uso:
 *   node scripts/export-lulo-mdb-csv.mjs "C:\Users\...\576PDVSA.MDB"
 *   npm run mdb:export-csv -- "C:\Users\...\576PDVSA.MDB" --out ./export_lulo_csv
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import MDBReader from 'mdb-reader';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const TABLAS_CRITICAS = new Set([
  'ObraApun',
  'ObraPart',
  'ObraMate',
  'ObraMano',
  'ObraEqui',
  'ObraApinMate',
  'ObraApinMano',
  'ObraApinEqui',
  'ObraCapi',
  'ObraCapiPart',
]);

function parseArgs(argv) {
  const flags = new Set();
  const pos = [];
  for (const a of argv) {
    if (a.startsWith('-')) flags.add(a);
    else pos.push(a);
  }
  const outIdx = argv.indexOf('--out');
  const outDir =
    outIdx >= 0 && argv[outIdx + 1] ? argv[outIdx + 1] : './export_lulo_csv';
  return {
    mdbPath: pos[0],
    outDir,
    soloCriticas: flags.has('--solo-criticas'),
  };
}

function escapeCsv(val) {
  if (val == null) return '';
  let s;
  if (val instanceof Date) s = val.toISOString();
  else if (Buffer.isBuffer(val)) s = val.toString('base64');
  else s = String(val);
  if (/[",\n\r]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function rowsToCsv(columns, rows) {
  const header = columns.map(escapeCsv).join(',');
  const lines = rows.map((row) => {
    const obj = row && typeof row === 'object' && !Array.isArray(row) ? row : {};
    return columns.map((col) => escapeCsv(obj[col])).join(',');
  });
  return `\ufeff${header}\n${lines.join('\n')}\n`;
}

function exportTable(reader, tableName, outDir) {
  const table = reader.getTable(tableName);
  const columns = table.getColumnNames();
  const rows = table.getData();
  const csv = rowsToCsv(columns, rows);
  const dest = path.join(outDir, `migracion_${tableName}.csv`);
  fs.writeFileSync(dest, csv, 'utf8');
  return rows.length;
}

function main() {
  const { mdbPath, outDir, soloCriticas } = parseArgs(process.argv.slice(2));
  if (!mdbPath) {
    console.error('❌ Indica la ruta al .mdb');
    console.error('   Ejemplo: npm run mdb:export-csv -- "C:\\...\\576PDVSA.MDB"');
    process.exit(1);
  }

  const abs = path.resolve(mdbPath);
  if (!fs.existsSync(abs)) {
    console.error(`❌ No existe: ${abs}`);
    process.exit(1);
  }

  const buffer = fs.readFileSync(abs);
  if (buffer.length < 2048) {
    console.error('❌ Archivo vacío o placeholder de iCloud. Usa el .mdb descargado completo.');
    process.exit(1);
  }

  let reader;
  try {
    reader = new MDBReader(buffer);
  } catch (e) {
    console.error('❌ Error al abrir MDB:', e instanceof Error ? e.message : e);
    process.exit(1);
  }

  const tablas = reader
    .getTableNames({ normalTables: true, systemTables: false })
    .sort((a, b) => a.localeCompare(b, undefined, { sensitivity: 'base' }));

  console.log('Tablas encontradas:', tablas.length);

  const objetivo = soloCriticas
    ? tablas.filter((t) => TABLAS_CRITICAS.has(t))
    : tablas.filter((t) => TABLAS_CRITICAS.has(t) || t.startsWith('Obra'));

  if (objetivo.length === 0) {
    console.error('❌ No hay tablas Obra* para exportar.');
    process.exit(1);
  }

  fs.mkdirSync(path.resolve(outDir), { recursive: true });

  for (const tabla of objetivo) {
    try {
      const filas = exportTable(reader, tabla, path.resolve(outDir));
      console.log(`Exportada: ${tabla} (${filas} filas)`);
    } catch (e) {
      console.warn(`⚠️  ${tabla}: ${e instanceof Error ? e.message : e}`);
    }
  }

  console.log(`\n✅ CSV en: ${path.resolve(outDir)}`);
}

main();

File: scripts/export-lulo-mdb-csv.py
#!/usr/bin/env python3
"""
Exporta tablas de un presupuesto LuloWin (.mdb / .accdb) a CSV.

Requisitos (Windows):
  pip install pandas pyodbc
  Driver: "Microsoft Access Database Engine" o "Microsoft Access Driver (*.mdb, *.accdb)"

Uso:
  python scripts/export-lulo-mdb-csv.py "C:\\ruta\\576PDVSA.MDB"
  python scripts/export-lulo-mdb-csv.py ./576PDVSA.MDB --out ./export_lulo
  python scripts/export-lulo-mdb-csv.py ./obra.mdb --solo-criticas
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

import pandas as pd
import pyodbc

TABLAS_CRITICAS = frozenset(
    {"ObraApun", "ObraPart", "ObraMate", "ObraMano", "ObraEqui"}
)


def conectar(mdb_path: Path) -> pyodbc.Connection:
    conn_str = (
        r"DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};"
        rf"DBQ={mdb_path.resolve()};"
    )
    return pyodbc.connect(conn_str)


def listar_tablas(conn: pyodbc.Connection) -> list[str]:
    cursor = conn.cursor()
    return sorted(
        t.table_name
        for t in cursor.tables(tableType="TABLE")
        if t.table_name and not t.table_name.startswith("MSys")
    )


def exportar_tabla(conn: pyodbc.Connection, tabla: str, out_dir: Path) -> int:
    query = f"SELECT * FROM [{tabla}]"
    df = pd.read_sql(query, conn)
    dest = out_dir / f"migracion_{tabla}.csv"
    df.to_csv(dest, index=False, encoding="utf-8-sig")
    return len(df)


def main() -> int:
    parser = argparse.ArgumentParser(description="Exportar MDB Lulo a CSV")
    parser.add_argument("mdb", type=Path, help="Ruta al archivo .mdb / .accdb")
    parser.add_argument(
        "--out",
        type=Path,
        default=Path("export_lulo_csv"),
        help="Carpeta de salida (default: ./export_lulo_csv)",
    )
    parser.add_argument(
        "--solo-criticas",
        action="store_true",
        help="Solo ObraApun, ObraPart, ObraMate, ObraMano, ObraEqui",
    )
    parser.add_argument(
        "--todas-obra",
        action="store_true",
        help="Todas las tablas que empiezan con Obra (default si no --solo-criticas)",
    )
    args = parser.parse_args()

    mdb = args.mdb
    if not mdb.is_file():
        print(f"Error: no existe el archivo {mdb}", file=sys.stderr)
        return 1

    out_dir = args.out
    out_dir.mkdir(parents=True, exist_ok=True)

    try:
        conn = conectar(mdb)
    except pyodbc.Error as e:
        print("Error de conexión. ¿Instalado el driver Access ODBC?", file=sys.stderr)
        print(e, file=sys.stderr)
        return 1

    try:
        tablas = listar_tablas(conn)
        print("Tablas encontradas:", ", ".join(tablas))

        if args.solo_criticas:
            objetivo = [t for t in tablas if t in TABLAS_CRITICAS]
        else:
            objetivo = [t for t in tablas if t in TABLAS_CRITICAS or t.startswith("Obra")]

        if not objetivo:
            print("No hay tablas Obra* para exportar.", file=sys.stderr)
            return 1

        for tabla in objetivo:
            filas = exportar_tabla(conn, tabla, out_dir)
            print(f"Exportada: {tabla} ({filas} filas) -> {out_dir / f'migracion_{tabla}.csv'}")

        print(f"\nListo. CSV en: {out_dir.resolve()}")
        return 0
    except Exception as e:
        print("Error en la extracción:", e, file=sys.stderr)
        return 1
    finally:
        conn.close()


if __name__ == "__main__":
    raise SystemExit(main())
